### DISCUSSION TITLE
Actions/cache - Fan in from matrix jobs

### DISCUSSION THREAD

**Discussion by nyarly** [community/community#26665] [Actions]
<p>I’m trying to build a Workflow to manage a relatively intensive build job. Specifically, I need to shard the build into manageable pieces, and then run the sharded build on self-hosted runners for memory capacity. Once builds are complete, I need to pull them into a merge job and continue with processing.</p>
<p>In a CI space, this should be a fan-out/fan-in pattern, and while those are generally tricky, I’m having particular trouble setting this up in Github Actions.</p>
<p>The shards are straightforward: I have a setup job that produces a matrix JSON, and then the individual builds are run in a strategy.</p>
<p>I’ve been trying to set up <code>actions/cache@v3</code> to store the build products and then pull them into the merge job. I’m currently doing something like <code>with: key: ${{ runner.os }}-cue-${{ hashFiles('**/*.cue') }}-${{ matrix.shard }}</code> in the builds, and then separate steps in the merge job, where the shard number is hard coded.</p>
<p>I have two problems:</p>
<ol>
<li>The merge job doesn’t see the cache from the matrix build jobs. This especially problematic because it’s impeding me sharing the output files of one job with successors.</li>
<li>I have to hard-code the cache steps; I’m hoping that I don’t need to write my own “multi-cache” job to be able to feed the matrix into and fetch everything, but that’s the conclusion I’m coming to.</li>
</ol>
<p>For debug logs related to <span class="hashtag">#1</span>, in the build job I’m seeing:</p>
<pre><code class="lang-auto">Cache not found for input keys: Linux-cue-96cc5790c3625d40c59224aac1d6bb80df6a79ba72dcf42462fb4d1dc8a109f9-test-1
...
##[debug]Resource Url: https://artifactcache.actions.githubusercontent.com/M0nxhPDVR3yfeu4g1I8mXeTsf6Ux3aND9UQaDXmbFNeioujIts/_apis/artifactcache/caches/6
Cache saved successfully
Cache saved with key: Linux-cue-96cc5790c3625d40c59224aac1d6bb80df6a79ba72dcf42462fb4d1dc8a109f9-test-1
</code></pre>
<p>and then in the merge job</p>
<pre><code class="lang-auto">##[debug]Resource Url: https://artifactcache.actions.githubusercontent.com/M0nxhPDVR3yfeu4g1I8mXeTsf6Ux3aND9UQaDXmbFNeioujIts/_apis/artifactcache/cache?keys=Linux-cue-96cc5790c3625d40c59224aac1d6bb80df6a79ba72dcf42462fb4d1dc8a109f9-test-1&amp;version=566590d5a3ae57874efff8ebebcdb0b21184f260230bf4ec8a4130daf459216f
Cache not found for input keys: Linux-cue-96cc5790c3625d40c59224aac1d6bb80df6a79ba72dcf42462fb4d1dc8a109f9-test-1
</code></pre>
<p>These are jobs running in the same workflow, so issues like cross-branch builds shouldn’t be an issue.</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>The cache is not meant for reliable transfer, it’s meant to avoid duplicate work on a best-effort basis. So I’d advise against this approach even if it happened to be working right now. :sweat_smile:</p>
<p>What you should do instead is to use artifacts, see <a href="https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts#passing-data-between-jobs-in-a-workflow" class="inline-onebox">Storing workflow data as artifacts - GitHub Docs</a>. Artifacts are meant exactly for the kind of scenario you describe, where data needs to be passed between parts of a build (or out of it).</p>
<p>If your artifacts are large and you don’t want to keep using the storage, you can <a href="https://github.com/actions/upload-artifact#retention-period">set a short retention period</a> when uploading the artifact.</p>

**Reply by dekimsey** [reply]
Thanks for the pointer, that's a good point. However, it doesn't answer how to perform the fan-in operation described by the OP. To the best of my knowledge `actions/download-action` has no mechanism to perform a download of artifacts matching some glob [#214](https://github.com/actions/download-artifact/issues/214).

So right now, there still isn't really any good way to be able to fan-in the results of a matrix job.

Note: `actions/download-action` can either download one artifact, or all artifacts. There is no in-between. For large build jobs downloading all artifacts can be very expensive. Especially if all one is trying to collate is the unit test results.
